nav2_map_server/src/map_saver.cpp

```diff
@@ -30,221 +31,173 @@
-    try {
-      map_mode = map_mode_from_string(mode_str);
-    } catch (std::invalid_argument &) {
-      map_mode = MapMode::Trinary;
-      RCLCPP_WARN(
-        get_logger(), "Map mode parameter not recognized: '%s', using default value (trinary)",
-        mode_str.c_str());
-    }
+  save_map_timeout_ = std::make_shared<rclcpp::Duration>(
+    std::chrono::milliseconds(declare_parameter("save_map_timeout", 2000)));
 
-    image_format = declare_parameter("image_format", map_mode == MapMode::Scale ? "png" : "pgm");
-    std::transform(
-      image_format.begin(), image_format.end(), image_format.begin(),
-      [](unsigned char c) {return std::tolower(c);});
-    const std::vector<std::string> BLESSED_FORMATS{"bmp", "pgm", "png"};
-    if (
-      std::find(BLESSED_FORMATS.begin(), BLESSED_FORMATS.end(), image_format) ==
-      BLESSED_FORMATS.end())
-    {
-      std::stringstream ss;
-      bool first = true;
-      for (auto & format_name : BLESSED_FORMATS) {
-        if (!first) {
-          ss << ", ";
+  free_thresh_default_ = declare_parameter("free_thresh_default", 25),
+  occupied_thresh_default_ = declare_parameter("occupied_thresh_default", 65);
+
+  map_listener_ = rclcpp::Node::make_shared("map_listener");
+}
+
+MapSaver::~MapSaver()
+{
+  RCLCPP_INFO(get_logger(), "Destroying");
+}
+
+nav2_util::CallbackReturn
+MapSaver::on_configure(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Configuring");
+  // Create SaveMap service callback handle
+  auto save_map_callback= [this](
+    const std::shared_ptr<rmw_request_id_t>/*request_header*/,
+    const std::shared_ptr<nav2_msgs::srv::SaveMap::Request> request,
+    std::shared_ptr<nav2_msgs::srv::SaveMap::Response> response) -> void {
+        // Set input arguments and call saveMapTopicToFile()
+        SaveParameters save_parameters;
+        save_parameters.map_file_name = request->map_url;
+        save_parameters.image_format = request->image_format;
+        save_parameters.free_thresh = request->free_thresh;
+        save_parameters.occupied_thresh = request->occupied_thresh;
+        try {
+          save_parameters.mode = map_mode_from_string(request->map_mode);
+        } catch (std::invalid_argument &) {
+          save_parameters.mode = MapMode::Trinary;
+          RCLCPP_WARN(
+            get_logger(), "Map mode parameter not recognized: '%s', using default value (trinary)",
+            request->map_mode.c_str());
         }
-        ss << "'" << format_name << "'";
-        first = false;
-      }
+
+        std::string map_topic = request->map_topic;
+
+	response->result = saveMapTopicToFile(map_topic, save_parameters);
+    };
+
+  // Make name prefix for services
+  std::string service_prefix = get_name() + std::string("/");
+
+  // Create a service that saves the occupancy grid from map topic to a file
+  save_map_service_ = create_service<nav2_msgs::srv::SaveMap>(
```

Comment:
This seems to be server specific code (e.g. not used by the CLI), this service should be put into the `map_saver_server` class. The lambda callback it calls should go there too, but the  `saveMapTopicToFile` function can stay here. That function can be used by the CLI and server. 

Perhaps it would be useful to go through this file and ask the question: "will any map saving application need this code?". if the answer is "yes" then leave it, if "no" then add to the appropriate CLI / server. 

It also does enough I think that `save_map_callback` should be a separate method as it doesn't use any local scope variables.

Reply:
> This seems to be server specific code (e.g. not used by the CLI), this service should be put into the map_saver_server class. The lambda callback it calls should go there too, but the saveMapTopicToFile function can stay here. That function can be used by the CLI and server.

> Perhaps it would be useful to go through this file and ask the question: "will any map saving application need this code?". if the answer is "yes" then leave it, if "no" then add to the appropriate CLI / server.

In the last change, `MapSaver` class is designed to work with both `map_saver_cli` and `map_saver_server` executables. There is no `MapSaverServer` class: `MapSaverServer` class was renamed to `MapSaver`; previous version of `MapSaver` was removed as unnecessary (all its functionality already moved to MapIO). `saveMapTopicToFile()` method is called directly both by CLI executable as well as from `save_map_callback` initiated by map saving service handled in `map_saver_server` executable. So, I suppose to remain it as is without separating `MapSaver` back to `MapSaver` and `MapSaverServer` classes.

Reply:
> It also does enough I think that save_map_callback should be a separate method as it doesn't use any local scope variables.

Agree. I think it is also reasonable for `handle_occ_callback()` and `load_map_callback()` from `MapServer` to make them separate methods.

Reply:
Done in https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-a42e5d09c9ddddb03729320e0be84f12R172, https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-a42e5d09c9ddddb03729320e0be84f12R188 and https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-99885e6562ca50edd8cc80db6362d317R94